diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x C:/Kuhring/SeqAn/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/fwmf9kkqo37idut8_stats.txt -x C:/Kuhring/SeqAn/misc/seqan_instrumentation/userdata/fwmf9kkqo37idut8_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x C:/Kuhring/SeqAn/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x C:/Kuhring/SeqAn/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x C:/Kuhring/SeqAn/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp ./sandbox/my_sandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 12:24:00.765625000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 12:27:10.156250000 +0200
@@ -44,6 +44,10 @@
 	Iterator<String<AminoAcid>, Standard>::Type it = begin(pep);
 	Iterator<String<AminoAcid>, Standard>::Type itEnd = end(pep);
 
+	typedef typename Size<TAlphabet>::Type TSize;
+	TSize alphSize = ValueSize<TAlphabet>::VALUE;
+	TSize[] sizeDist = TSize[alphSize];
+
 	cout << pep << endl;
 	while (it != itEnd){
 		if (*it == 'R'){
@@ -51,5 +55,5 @@
 		}
 		it++;
 	}
-	cout << pep;
+	cout << pep << endl;
 }
