diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/xfdtznywb7qjx61s_stats.txt -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata/xfdtznywb7qjx61s_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp ./sandbox/my_sandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 14:28:35.000000000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 14:36:38.000000000 +0200
@@ -5,49 +5,10 @@
 
 using namespace seqan;
 
-template <typename TAlphabet>
-void showAllLetterOfMyAlphabet(TAlphabet const &)
-{
-    typedef typename Size<TAlphabet>::Type TSize;
-    TSize alphSize = ValueSize<TAlphabet>::VALUE;
-    for (TSize i = 0; i < alphSize; ++i)
-        std::cout << i << ',' << TAlphabet(i) << "  ";
-    std::cout << std::endl;
-}
-
-//template <typename String> class <CharString, DnaString, Peptide>
-template <typename TString>
-void countOneMers(TString const sequence)
-{
-	typedef typename Size<TString>::Type TSize;
-	typedef typename Value<TString>::Type TAlphabet;
-	
-	typedef String<TSize> TCounterString;
-	TSize alphSize = ValueSize<TAlphabet>::VALUE;
-	TCounterString table;
-	resize(table, alphSize,0); 
-	
-	typedef typename Iterator<TString>::Type TIter;
-    TIter itEnd = end(sequence);
-    for (TIter it = begin(sequence); it != itEnd; goNext(it))
-        value(table, ordValue(value(it))) += 1;
-	
-	
-	typedef typename Iterator<TCounterString>::Type TTableIter;	
-	TTableIter countIt = begin(table);
-    TTableIter countItEnd = end(table);
-	for(TSize pos=0; countIt != countItEnd; ++countIt, ++pos)
-	{
-		if(value(countIt > 0))
-		{
-			std::cout <<TAlphabet(pos) << ':' << value(countIt) << std::endl;
-		}
-	}
-
-}
-
+//template <typename TAlphabet>
 typedef String<char> TSequence;                 // sequence type
 template <typename TSequence>
+
 void align_sequences(TSequence & seqa, TSequence & seqb)
 {
 //	TSequence seqa = "aa";
diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/xfdtznywb7qjx61s_stats.txt -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata/xfdtznywb7qjx61s_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/old.cpp ./sandbox/my_sandbox/apps/my_app/old.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/old.cpp	1970-01-01 01:00:00.000000000 +0100
+++ ./sandbox/my_sandbox/apps/my_app/old.cpp	2011-09-13 14:29:59.000000000 +0200
@@ -0,0 +1,11 @@
+/*
+ *  old.cpp
+ *  seqan
+ *
+ *  Created by Roland Krause on 13.09.11.
+ *  Copyright 2011 MPI for Molecular Genetics. All rights reserved.
+ *
+ */
+
+#include "old.h"
+
diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/xfdtznywb7qjx61s_stats.txt -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata/xfdtznywb7qjx61s_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/old.h ./sandbox/my_sandbox/apps/my_app/old.h
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/old.h	1970-01-01 01:00:00.000000000 +0100
+++ ./sandbox/my_sandbox/apps/my_app/old.h	2011-09-13 14:36:14.000000000 +0200
@@ -0,0 +1,49 @@
+/*
+ *  old.h
+ *  seqan
+ *
+ *  Created by Roland Krause on 13.09.11.
+ *  Copyright 2011 MPI for Molecular Genetics. All rights reserved.
+ *
+ */
+
+void showAllLetterOfMyAlphabet(TAlphabet const &)
+{
+    typedef typename Size<TAlphabet>::Type TSize;
+    TSize alphSize = ValueSize<TAlphabet>::VALUE;
+    for (TSize i = 0; i < alphSize; ++i)
+        std::cout << i << ',' << TAlphabet(i) << "  ";
+    std::cout << std::endl;
+}
+
+//template <typename String> class <CharString, DnaString, Peptide>
+template <typename TString>
+void countOneMers(TString const sequence)
+{
+	typedef typename Size<TString>::Type TSize;
+	typedef typename Value<TString>::Type TAlphabet;
+	
+	typedef String<TSize> TCounterString;
+	TSize alphSize = ValueSize<TAlphabet>::VALUE;
+	TCounterString table;
+	resize(table, alphSize,0); 
+	
+	typedef typename Iterator<TString>::Type TIter;
+    TIter itEnd = end(sequence);
+    for (TIter it = begin(sequence); it != itEnd; goNext(it))
+        value(table, ordValue(value(it))) += 1;
+	
+	
+	typedef typename Iterator<TCounterString>::Type TTableIter;	
+	TTableIter countIt = begin(table);
+    TTableIter countItEnd = end(table);
+	for(TSize pos=0; countIt != countItEnd; ++countIt, ++pos)
+	{
+		if(value(countIt > 0))
+		{
+			std::cout <<TAlphabet(pos) << ':' << value(countIt) << std::endl;
+		}
+	}
+	
+}
+
