diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/xfdtznywb7qjx61s_stats.txt -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata/xfdtznywb7qjx61s_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp ./sandbox/my_sandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 11:51:13.000000000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 12:00:41.000000000 +0200
@@ -15,8 +15,24 @@
 template <typename CharString >
 void countOneMers(CharString str)
 {
+	typedef Size<CharString>::Type TSize;
+    typedef CharString<TSize> TCounterString;
+    TCounterString counter;
+	Iterator<String>::Type it = begin(str);
+	Iterator<String>::Type itEnd = end(str);
 	::std::cout << str << ::std::endl;
-//	Iterator<String<char> >::Type it = begin();
+	while(it != itEnd)
+	{
+		if(value(it) == 'R')
+		{
+			std::cout<< 'A';
+		} else {
+			std::cout << *it;
+		}
+		++it;
+		
+	}
+
 	
 }
 
