diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/xfdtznywb7qjx61s_stats.txt -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata/xfdtznywb7qjx61s_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /Users/krause/Workspace/seqan-trunk/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp ./sandbox/my_sandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 10:19:25.000000000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 10:23:42.000000000 +0200
@@ -1,11 +1,22 @@
+#include <seqan/sequence.h>
+#include <seqan/basic.h>
 #include <iostream>
-#include <seqan/sequence.h>  // CharString, ...
-#include <seqan/file.h>      // to stream a CharString into cout
+using namespace seqan;
 
-int main(int, char **) {
-    std::cout << "Hello World!" << std::endl;
-    seqan::CharString mySeqanString = "Hello SeqAn!";
-    std::cout << mySeqanString << std::endl;
-    return 1;
+template <typename TAlphabet>
+void showAllLetterOfMyAlphabet(TAlphabet const &)
+{
+    typedef typename Size<TAlphabet>::Type TSize;
+    TSize alphSize = ValueSize<TAlphabet>::VALUE;
+    for (TSize i = 0; i < alphSize; ++i)
+        std::cout << i << ',' << TAlphabet(i) << "  ";
+    std::cout << std::endl;
 }
 
+int main()
+{
+    showAllLetterOfMyAlphabet(AminoAcid());
+    showAllLetterOfMyAlphabet(Dna());
+    showAllLetterOfMyAlphabet(Dna5());
+    return 0;
+}
\ No newline at end of file
