diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.h ./sandbox/my_sandbox/apps/my_app/my_app.h
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.h	2011-09-13 10:33:27.000000000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.h	2011-09-13 10:34:19.000000000 +0200
@@ -122,11 +122,11 @@
     for (TIterator it = begin(options.texts); it != end(options.texts); ++it) {
         std::cout << "  " << *it << std::endl;
     }
-    TIterator it=begin(options.texts)
+    TIterator it=begin(options.texts);
 		CharString str1=(*it);
 		File<Sync<> > myFile1;
 		if (!open(myFile1, str1, OPEN_WRONLY | OPEN_CREATE)) {
-			cout << "Could not open for writing file " << str1 << endl;
+			std::cout << "Could not open for writing file " << str1 << endl;
 			return 1;
 		}
 		++it;
diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.h~ ./sandbox/my_sandbox/apps/my_app/my_app.h~
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.h~	2011-09-13 10:33:27.000000000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.h~	2011-09-13 10:34:19.000000000 +0200
@@ -77,13 +77,13 @@
 void
 setupCommandLineParser(CommandLineParser & parser, Options const & options)
 {
-    addVersionLine(parser, "0.1");
+    addVersionLine(parser, "0.0.1");
     
     addTitleLine(parser, "**********************");
     addTitleLine(parser, "* my_app *");
     addTitleLine(parser, "**********************");
     addTitleLine(parser, "");
-    addTitleLine(parser, "(c) 2011 by Your Name <your.email@example.net>");
+    addTitleLine(parser, "(c) 2011 by my Name <my.email@example.net>");
 
     addUsageLine(parser, "[OPTIONS] TEXT+");
     
@@ -122,7 +122,18 @@
     for (TIterator it = begin(options.texts); it != end(options.texts); ++it) {
         std::cout << "  " << *it << std::endl;
     }
-    
+    TIterator it=begin(options.texts)
+		CharString str1=(*it);
+		File<Sync<> > myFile1;
+		if (!open(myFile1, str1, OPEN_WRONLY | OPEN_CREATE)) {
+			cout << "Could not open for writing file " << str1 << endl;
+			return 1;
+		}
+		++it;
+		CharString block1=(*it);
+		write(myFile1, block1, sizeof(block1));
+		close(myFile1);
+		
     return 0;
 }
 
