diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp ./sandbox/my_sandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-14 14:20:44.000000000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-14 14:21:37.000000000 +0200
@@ -115,7 +115,7 @@
 void globalAlign (String<T> str1, String<T> str2) {
 		typedef String<T> TSequence;
 		typedef Align<TSequence, ArrayGaps> TAlign;
- 		typedef Value<Row<TAlign> >::Type TRow;
+ 		typedef Row<Value<TAlign>::Type>::Type TRow;
 		TAlign align;
 		resize(rows(align),2);
 		assignSource(row(align,0),str1);
diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp~ ./sandbox/my_sandbox/apps/my_app/my_app.cpp~
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp~	2011-09-14 14:20:44.000000000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp~	2011-09-14 14:21:37.000000000 +0200
@@ -115,7 +115,7 @@
 void globalAlign (String<T> str1, String<T> str2) {
 		typedef String<T> TSequence;
 		typedef Align<TSequence, ArrayGaps> TAlign;
- 		typedef Row<TAlign>::Type TRow;
+ 		typedef Value<Row<TAlign> >::Type TRow;
 		TAlign align;
 		resize(rows(align),2);
 		assignSource(row(align,0),str1);
@@ -279,7 +279,7 @@
 // 		tut Alignments
 		typedef String<char> TSequence;
 		typedef Align<TSequence, ArrayGaps> TAlign;
-		typedef Value<Row<TAlign> >::Type TRow;
+		typedef Row<TAlign>::Type TRow;
 		TSequence seq1="CDFGHC", seq2="CDEFGAHC";
 		TAlign align;
 		resize(rows(align),2);
