diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp ./sandbox/my_sandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 12:30:05.000000000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 12:30:47.000000000 +0200
@@ -124,7 +124,7 @@
 				Tend=sysTime();
 				t_a+=Tend-Tstart;
 			}
-			std::cout << length(myString_arr);
+			std::cout << sizeof(myString_arr);
 				Tstart=sysTime();
 				clear (mySimpleAlloc);
 				Tend=sysTime();
@@ -143,7 +143,7 @@
 				Tend=sysTime();
 				t_a+=Tend-Tstart;
 			}
-			std::cout << length(myString_arr);
+			std::cout << sizeof(myString_arr);
 				Tstart=sysTime();
 				clear (myMultiPoolAlloc);
 				Tend=sysTime();
diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp~ ./sandbox/my_sandbox/apps/my_app/my_app.cpp~
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp~	2011-09-13 12:30:05.000000000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp~	2011-09-13 12:30:47.000000000 +0200
@@ -124,6 +124,7 @@
 				Tend=sysTime();
 				t_a+=Tend-Tstart;
 			}
+			std::cout << length(myString_arr);
 				Tstart=sysTime();
 				clear (mySimpleAlloc);
 				Tend=sysTime();
@@ -142,6 +143,7 @@
 				Tend=sysTime();
 				t_a+=Tend-Tstart;
 			}
+			std::cout << length(myString_arr);
 				Tstart=sysTime();
 				clear (myMultiPoolAlloc);
 				Tend=sysTime();
