diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp ./sandbox/my_sandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-14 14:29:07.000000000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-14 14:29:45.000000000 +0200
@@ -114,8 +114,8 @@
 template <typename T>
 void globalAlign (String<T> str1, String<T> str2) {
 		typedef String<T> TSequence;
-		typedef Align<TSequence, ArrayGaps>::Type TAlign;
- 		typedef Row<Align<String<T>, ArrayGaps> >::Type TRow;
+		typedef Align<TSequence, ArrayGaps> TAlign;
+ 		typedef Row<TAlign> TRow;
 		TAlign align;
 		resize(rows(align),2);
 		assignSource(row(align,0),str1);
diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata/fj1wwg5keyf1mh0e_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /home/tba/projects/SeqAn/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp~ ./sandbox/my_sandbox/apps/my_app/my_app.cpp~
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp~	2011-09-14 14:29:07.000000000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp~	2011-09-14 14:29:45.000000000 +0200
@@ -114,7 +114,7 @@
 template <typename T>
 void globalAlign (String<T> str1, String<T> str2) {
 		typedef String<T> TSequence;
-		typedef Align<TSequence, ArrayGaps> TAlign;
+		typedef Align<TSequence, ArrayGaps>::Type TAlign;
  		typedef Row<Align<String<T>, ArrayGaps> >::Type TRow;
 		TAlign align;
 		resize(rows(align),2);
@@ -127,9 +127,6 @@
 		int score = globalAlignment(align,Score<int>(1,-1,-1,-1));
 		std::cout << "Score = " << score << std::endl << align;
 		
-		for (int i=0; i<rows(align); i++) {
-			
-		}
 }
 
 // Program entry point
