diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/userdata/id.txt' -x misc/seqan_instrumentation/userdata/o6lmo5tpxvn3b6fg_stats.txt -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/userdata/o6lmo5tpxvn3b6fg_stats.txt' -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/bin' -x misc/seqan_instrumentation/last_revision_copy -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/last_revision_copy' -x misc/seqan_instrumentation/userdata -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/userdata' ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp ./sandbox/my_sandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 11:28:33.750000000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 11:36:35.062500000 +0200
@@ -4,15 +4,21 @@
 
 typedef seqan::String<seqan::Dna> DnaString;
 
-template <class A>
-int count1mers(typename seqan::String<A> str)
+int count1mers(seqan::String<char> str)
 {
-	int alphabetsize = seqan::ValueSize<A>::VALUE;
 	seqan::String<int> charcount;
-	seqan::resize(charcount, alphabetsize, 0);
+	seqan::resize(charcount, 256, 0);
 
-	for(int i = 0; )
+	for(int i = 0; sizeof(str); i++)
+	{
+		charcount[char(str[i])] ++;
+	}
 
+	for(int i = 0; sizeof(charcount); i++)
+	{
+		if(charcount[i] != 0)
+			std::cout << char(i) << " " << charcount[i] << std::endl;
+	}
 	return 1;
 }
 
