diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/userdata/id.txt' -x misc/seqan_instrumentation/userdata/o6lmo5tpxvn3b6fg_stats.txt -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/userdata/o6lmo5tpxvn3b6fg_stats.txt' -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/bin' -x misc/seqan_instrumentation/last_revision_copy -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/last_revision_copy' -x misc/seqan_instrumentation/userdata -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/userdata' ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp ./sandbox/my_sandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 12:07:03.828125000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 12:10:08.531250000 +0200
@@ -4,6 +4,8 @@
 #include <seqan/file.h>      // to stream a CharString into cout
 
 typedef seqan::String<seqan::Dna> DnaString;
+typedef seqan::String<seqan::Peptide> PeptideString;
+typedef seqan::String<char> CharString;
 
 template <typename A>
 int count1mers(seqan::String<A> str)
@@ -27,5 +29,13 @@
 	DnaString testseq = "CAGTGTGAGCGATTACATACCATCGACTAGT";
     std::cout << testseq << std::endl;
 	count1mers(testseq);
+
+	PeptideString testpep = "MQDRVKRPMNAFIVWSRDQRRKMALEN";
+    std::cout << testpep << std::endl;
+	count1mers(testpep);
+
+	CharString testchar = "Hello World!";
+    std::cout << testchar << std::endl;
+	count1mers(testchar);
     return 1;
 }
\ No newline at end of file
