diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/userdata/id.txt' -x misc/seqan_instrumentation/userdata/o6lmo5tpxvn3b6fg_stats.txt -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/userdata/o6lmo5tpxvn3b6fg_stats.txt' -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/bin' -x misc/seqan_instrumentation/last_revision_copy -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/last_revision_copy' -x misc/seqan_instrumentation/userdata -x 'C:/Dokumente und Einstellungen/lindnerm/Desktop/seqan/misc/seqan_instrumentation/userdata' ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp ./sandbox/my_sandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 15:19:59.203125000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 15:26:09.765625000 +0200
@@ -73,6 +73,17 @@
 	seqan::assignSource(seqan::row(align,1),s2);
 	int score = seqan::globalAlignment(align,seqan::Score<int>(1,-1,-1,-1), seqan::Hirschberg());
 
+	std::cout << "Alignmnent score: " << score << std::endl;
+
+	auto it1 = seqan::begin(seqan::row(align,0));
+	auto end1 = seqan::end(seqan::row(align,0));
+	auto it2 = seqan::begin(seqan::row(align,1));
+	auto end2 = seqan::end(seqan::row(align,1));
+
+	for(; it1 != end1; it1++)
+		std::cout << *it1;
+	std::cout << std::endl;
+
 	
 }
 
