diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x D:/workspaceCPP/eclipse/seqan_workshop/seqan/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/20xjp6c6gjzthtye_stats.txt -x D:/workspaceCPP/eclipse/seqan_workshop/seqan/misc/seqan_instrumentation/userdata/20xjp6c6gjzthtye_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x D:/workspaceCPP/eclipse/seqan_workshop/seqan/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x D:/workspaceCPP/eclipse/seqan_workshop/seqan/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x D:/workspaceCPP/eclipse/seqan_workshop/seqan/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/mySandbox/apps/my_app/my_app.cpp ./sandbox/mySandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/mySandbox/apps/my_app/my_app.cpp	2011-09-13 08:39:25.993602700 -0500
+++ ./sandbox/mySandbox/apps/my_app/my_app.cpp	2011-09-13 08:40:22.340825600 -0500
@@ -20,7 +20,7 @@
 	TIter itrStart = begin(AA);
 	TIter itrEnd = end(AA);
 	for(;itrStart!=itrEnd; ++itrStart){
-		cout << ", " << *itrStart << " ";
+		cout << *itrStart << endl;
 		if(*itrStart == 'R')
 			*itrStart = 'A';
 	}
