diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /home/thor/workspace/SeqAn/SeqAn/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/fec6t5zk4j2ik4e7_stats.txt -x /home/thor/workspace/SeqAn/SeqAn/misc/seqan_instrumentation/userdata/fec6t5zk4j2ik4e7_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /home/thor/workspace/SeqAn/SeqAn/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /home/thor/workspace/SeqAn/SeqAn/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /home/thor/workspace/SeqAn/SeqAn/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp ./sandbox/my_sandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 15:12:46.000000000 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 15:13:33.654994677 +0200
@@ -64,15 +64,16 @@
 
 	it = begin(aaString);
 
-	while (!atEnd(it)) {
+	while (it != last) {
 		value(counterString, ordValue(value(it))) += 1;
 		it++;
 	}
 
 
 	Iterator<String<TSize> >::Type it2 = begin(counterString);
+	Iterator<String<TSize> >::Type last2 = end(counterString);
 	TSize pos = 0;
-	while (!atEnd(it2)) {
+	while (it2 != last2) {
 		std::cout << AminoAcid(pos++) << " " << value(it2) << std::endl;
 		it2++;
 	}
