diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x Z:/RAID/Seqan/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/gxymtwhk6d1fehg3_stats.txt -x Z:/RAID/Seqan/misc/seqan_instrumentation/userdata/gxymtwhk6d1fehg3_stats.txt -x .svn -x bin -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x Z:/RAID/Seqan/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x Z:/RAID/Seqan/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/last_revision_copy -x Z:/RAID/Seqan/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x Z:/RAID/Seqan/misc/seqan_instrumentation/userdata -x misc/seqan_instrumentation/userdata -x Z:/RAID/Seqan/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/wissmann/demos/task2.cpp ./sandbox/wissmann/demos/task2.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/wissmann/demos/task2.cpp	2012-04-12 12:41:47.968205000 +0200
+++ ./sandbox/wissmann/demos/task2.cpp	2012-04-13 16:09:55.189566000 +0200
@@ -158,7 +158,7 @@
 	assignSource (row(align4,1), Seq6);
 
 	LocalAlignmentFinder<> finder(align4);
-
+//
 	Score<int> scoring (3,-2,-1,-5);
 
 	for (int i = 0; i<3 ; i++)
