diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /home/felix/BioStore/seqan-trunk/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/u35c5glxw45elmls_stats.txt -x /home/felix/BioStore/seqan-trunk/misc/seqan_instrumentation/userdata/u35c5glxw45elmls_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /home/felix/BioStore/seqan-trunk/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /home/felix/BioStore/seqan-trunk/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /home/felix/BioStore/seqan-trunk/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/ministellar/ministellar.cpp ./sandbox/my_sandbox/apps/ministellar/ministellar.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/ministellar/ministellar.cpp	2011-09-13 12:48:31.000000000 +0200
+++ ./sandbox/my_sandbox/apps/ministellar/ministellar.cpp	2011-09-13 12:51:37.525727295 +0200
@@ -118,9 +118,9 @@
 	//       Check that the extended seed has a length of at least minimalLength.
 	// HINT: Have a look at the Seed-and-Extend tutorial.
         typedef Seed<Simple> TSeed;
-        TSeed seed = TSeed(beginPosition(infix1) + clippedBeginPosition(row(aln,0)),
+        TSeed seed = TSeed(beginPosition(infix1) + clippedBeginPosition(row(aln, 0)),
+                           beginPosition(infix2) + clippedBeginPosition(row(aln, 1)),
                            beginPosition(infix1) + clippedEndPosition(row(aln, 0)),
-                           beginPosition(infix2) + clippedBeginPosition(row(aln,1)),
                            beginPosition(infix2) + clippedEndPosition(row(aln, 1)));
 
        extendSeed(seed, xDrop, scoring, seqs1[0], seqs2[0], 2, GappedXDrop());
