diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x C:/Users/peter/Documents/seqan/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/0meio6dzt3eo1wj7_stats.txt -x C:/Users/peter/Documents/seqan/misc/seqan_instrumentation/userdata/0meio6dzt3eo1wj7_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x C:/Users/peter/Documents/seqan/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x C:/Users/peter/Documents/seqan/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x C:/Users/peter/Documents/seqan/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp ./sandbox/my_sandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 12:10:42.541992100 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 12:11:07.344726500 +0200
@@ -60,7 +60,7 @@
 
 	//PeptideIterator it1 = begin(str);
 
-	//absFreqOfLettersInSeq(it1);
+	absFreqOfLettersInSeq(it1);
 
 //	showAllLetterOfMyAlphabet(AminoAcid());
 //   showAllLetterOfMyAlphabet(Dna());
