diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x C:/Users/peter/Documents/seqan/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/0meio6dzt3eo1wj7_stats.txt -x C:/Users/peter/Documents/seqan/misc/seqan_instrumentation/userdata/0meio6dzt3eo1wj7_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x C:/Users/peter/Documents/seqan/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x C:/Users/peter/Documents/seqan/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x C:/Users/peter/Documents/seqan/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp ./sandbox/my_sandbox/apps/my_app/my_app.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 12:24:21.829101500 +0200
+++ ./sandbox/my_sandbox/apps/my_app/my_app.cpp	2011-09-13 12:42:16.644531200 +0200
@@ -53,17 +53,28 @@
 
 int main(int, char **)
 {
-
-	String<AminoAcid> str = "MQDRVKRPMNAFIVWSRDQRRKMALEN";
-
-    typedef Iterator<TAminoAcidString>::Type TIter;
-    TIter itEnd = end(str);
-    for (TIter it = begin(str); it != itEnd; goNext(it))
-    {
-        if (value(it) == 'R') value(it) = 'A';
-        std::cout << value(it) << ',';
-    }
-    std::cout << std::endl;
+	seqan::String<char> str = "admn";
+	seqan::Iterator<seqan::String<char> >::Type it = begin(str);
+	seqan::Iterator<seqan::String<char> >::Type itEnd = end(str);
+
+	while (it != itEnd)
+	{
+		::std::cout << *it;
+		++it;
+	}
+	::std::cout << ::std::endl;
+
+		
+	//String<AminoAcid> str = "MQDRVKRPMNAFIVWSRDQRRKMALEN";
+
+    //typedef Iterator<TAminoAcidString>::Type TIter;
+    //((TIter itEnd = end(str);
+    //for (TIter it = begin(str); it != itEnd; goNext(it))
+    //{
+      //  if (value(it) == 'R') value(it) = 'A';
+        //std::cout << value(it) << ',';
+    //}
+    //std::cout << std::endl;
 
 	//PeptideIterator it1 = begin(str);
 
