diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x '/home/paul/Projects/2011 SeqAn Workshop/seqan/misc/seqan_instrumentation/userdata/id.txt' -x misc/seqan_instrumentation/userdata/amudto8y1mzxaebv_stats.txt -x '/home/paul/Projects/2011 SeqAn Workshop/seqan/misc/seqan_instrumentation/userdata/amudto8y1mzxaebv_stats.txt' -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x '/home/paul/Projects/2011 SeqAn Workshop/seqan/misc/seqan_instrumentation/bin' -x misc/seqan_instrumentation/last_revision_copy -x '/home/paul/Projects/2011 SeqAn Workshop/seqan/misc/seqan_instrumentation/last_revision_copy' -x misc/seqan_instrumentation/userdata -x '/home/paul/Projects/2011 SeqAn Workshop/seqan/misc/seqan_instrumentation/userdata' ./misc/seqan_instrumentation/last_revision_copy/sandbox/mordor/apps/exastellar/exastellar.cpp ./sandbox/mordor/apps/exastellar/exastellar.cpp
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/mordor/apps/exastellar/exastellar.cpp	2011-09-13 12:02:19.000000000 +0200
+++ ./sandbox/mordor/apps/exastellar/exastellar.cpp	2011-09-13 12:03:15.998735297 +0200
@@ -19,6 +19,16 @@
 TScoreValue const minScore = 9;
 TScoreValue const xDrop = 3;
 
+template<typename TSeed, typename TSeq>
+void
+writeSeed(TSeed & seed, TSeq const & seq0, TSeq const & seq1) {
+    std::cout << "Seed from position " << leftPosition(seed, 0);
+    std::cout << " to " << rightPosition(seed, 0) << ": ";
+    std::cout << infix(seq0, leftPosition(seed, 0), rightPosition(seed, 0)+1) << std::endl;
+    std::cout << "Seed from position " << leftPosition(seed, 1);
+    std::cout << " to " << rightPosition(seed, 1) << ": ";
+    std::cout << infix(seq1, leftPosition(seed, 1), rightPosition(seed, 1)+1) << std::endl;
+}
 
 int main(int argc, char const ** argv)
 {
diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x '/home/paul/Projects/2011 SeqAn Workshop/seqan/misc/seqan_instrumentation/userdata/id.txt' -x misc/seqan_instrumentation/userdata/amudto8y1mzxaebv_stats.txt -x '/home/paul/Projects/2011 SeqAn Workshop/seqan/misc/seqan_instrumentation/userdata/amudto8y1mzxaebv_stats.txt' -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x '/home/paul/Projects/2011 SeqAn Workshop/seqan/misc/seqan_instrumentation/bin' -x misc/seqan_instrumentation/last_revision_copy -x '/home/paul/Projects/2011 SeqAn Workshop/seqan/misc/seqan_instrumentation/last_revision_copy' -x misc/seqan_instrumentation/userdata -x '/home/paul/Projects/2011 SeqAn Workshop/seqan/misc/seqan_instrumentation/userdata' ./misc/seqan_instrumentation/last_revision_copy/sandbox/mordor/apps/exastellar/exastellar.cpp~ ./sandbox/mordor/apps/exastellar/exastellar.cpp~
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/mordor/apps/exastellar/exastellar.cpp~	2011-09-13 11:59:18.000000000 +0200
+++ ./sandbox/mordor/apps/exastellar/exastellar.cpp~	2011-09-13 12:02:19.568735312 +0200
@@ -92,12 +92,11 @@
 	        return -1;
 	    }
 	    
-	    SeedSet< int, Seed< Simple >, Score< TScoreValue >, Scored > wind;
-	    addSeed( wind, clippedBeginPosition( row( the_showdown, 0 ) ), clippedBeginPosition( row( the_showdown, 1 ) ), clippedEndPosition( row( the_showdown, 0 ) ), clippedEndPosition( row( the_showdown, 1 ) ) );
-	    writeSeed( wind[0], seq1, seq2 );	    	    
-	    extendSeed( wind[0], xDrop, judgement_of_carrion, seq1, seq2, 2, GappedXDrop() );
+	    Seed< Simple > wind( clippedBeginPosition( row( the_showdown, 0 ) ), clippedBeginPosition( row( the_showdown, 1 ) ), clippedEndPosition( row( the_showdown, 0 ) ), clippedEndPosition( row( the_showdown, 1 ) ) );
+	    writeSeed( wind, seq1, seq2 );	    	    
+	    extendSeed( wind, xDrop, judgement_of_carrion, seq1, seq2, 2, GappedXDrop() );
 	    std::cout << "After extension:" << std::endl;
-	    writeSeed( wind[0], seq1, seq2 );
+	    writeSeed( wind, seq1, seq2 );
 	    std::cout << "---" << std::endl;
 	}	
 
