diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/y95d44txbnpnrnz0_stats.txt -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/userdata/y95d44txbnpnrnz0_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/CMakeLists.txt ./sandbox/my_sandbox/CMakeLists.txt
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/CMakeLists.txt	1970-01-01 01:00:00.000000000 +0100
+++ ./sandbox/my_sandbox/CMakeLists.txt	2011-09-13 09:59:49.000000000 +0200
@@ -0,0 +1,11 @@
+cmake_minimum_required (VERSION 2.6)
+project (seqan_sandbox_my_sandbox)
+
+# We have to register our own include directory here, though.
+seqan_setup_includes(include SeqAnSandboxMy_sandbox)
+# SeqAnCore is added include in ../core/CMakeLists.txt.  We only have to
+# register that we want it here, too.  Same for SeqAnExtras.
+seqan_make_seqan_available(SeqAnCore)
+seqan_make_seqan_available(SeqAnExtras)
+
+seqan_add_all_subdirectories()  # include_directory(*)
diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/y95d44txbnpnrnz0_stats.txt -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/userdata/y95d44txbnpnrnz0_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/CMakeLists.txt ./sandbox/my_sandbox/apps/CMakeLists.txt
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/apps/CMakeLists.txt	1970-01-01 01:00:00.000000000 +0100
+++ ./sandbox/my_sandbox/apps/CMakeLists.txt	2011-09-13 09:59:49.000000000 +0200
@@ -0,0 +1,6 @@
+cmake_minimum_required (VERSION 2.6)
+project (seqan_sandbox_my_sandbox_apps)
+
+seqan_setup_apps(sandbox_my_sandbox_apps)  # Add meta target and make executables depend on it.
+
+seqan_add_all_subdirectories()  # include_directory(*)
diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/y95d44txbnpnrnz0_stats.txt -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/userdata/y95d44txbnpnrnz0_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/demos/CMakeLists.txt ./sandbox/my_sandbox/demos/CMakeLists.txt
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/demos/CMakeLists.txt	1970-01-01 01:00:00.000000000 +0100
+++ ./sandbox/my_sandbox/demos/CMakeLists.txt	2011-09-13 09:59:49.000000000 +0200
@@ -0,0 +1,6 @@
+cmake_minimum_required (VERSION 2.6)
+project (seqan_sandbox_my_sandbox_demos)
+
+seqan_setup_demos(sandbox_my_sandbox_demos)  # Add meta target and make executables depend on it.
+
+seqan_add_all_executables(sandbox_my_sandbox_demos)
diff -u -r -N -x '*.o' -x Thumbs.db -x .DS_Store -x CMakeCache.txt -x misc/seqan_instrumentation/userdata/id.txt -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/userdata/id.txt -x misc/seqan_instrumentation/userdata/y95d44txbnpnrnz0_stats.txt -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/userdata/y95d44txbnpnrnz0_stats.txt -x .svn -x build -x util -x misc -x docs -x misc/seqan_instrumentation/bin -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/bin -x misc/seqan_instrumentation/last_revision_copy -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/last_revision_copy -x misc/seqan_instrumentation/userdata -x /Applications/GENOME_ANALYSIS/seqan-trunk/misc/seqan_instrumentation/userdata ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/tests/CMakeLists.txt ./sandbox/my_sandbox/tests/CMakeLists.txt
--- ./misc/seqan_instrumentation/last_revision_copy/sandbox/my_sandbox/tests/CMakeLists.txt	1970-01-01 01:00:00.000000000 +0100
+++ ./sandbox/my_sandbox/tests/CMakeLists.txt	2011-09-13 09:59:49.000000000 +0200
@@ -0,0 +1,6 @@
+cmake_minimum_required (VERSION 2.6)
+project (seqan_sandbox_my_sandbox_tests)
+
+seqan_setup_tests(sandbox_my_sandbox_tests)  # Setup flags etc, add meta target.
+
+seqan_add_all_subdirectories()  # include_directory(*)
