<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.3.1"/>
<title>SeqDPT: D:/SeqAn/Development/seqan-trunk/sandbox/my_sandbox/apps/SeqDPT/adapterTrimming.h File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectlogo"><img alt="Logo" src="seqan_logo_167x110[1].png"/></td>
  <td style="padding-left: 0.5em;">
   <div id="projectname">SeqDPT
   </div>
   <div id="projectbrief">NGS Data Postprogressing Toolbox</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.3.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="dir_17781b3159a5fccb1c42569687fc4d9b.html">Development</a></li><li class="navelem"><a class="el" href="dir_8189ab751ded8ac4cea9718ea95de3b7.html">seqan-trunk</a></li><li class="navelem"><a class="el" href="dir_e6afdf0727609f4c22bfa8fdca52c8fc.html">sandbox</a></li><li class="navelem"><a class="el" href="dir_fc37661ed34503c9e24dd8c3c73ad211.html">my_sandbox</a></li><li class="navelem"><a class="el" href="dir_33a4b5dbfed2050c7893d2d8375d6565.html">apps</a></li><li class="navelem"><a class="el" href="dir_03f83222c96366803fc0f385d16c833f.html">SeqDPT</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#typedef-members">Typedefs</a> &#124;
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<div class="title">adapterTrimming.h File Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Contains the functions for adapter trimming.  
<a href="#details">More...</a></p>
<div class="textblock"><code>#include &lt;seqan/align.h&gt;</code><br/>
<code>#include &lt;seqan/find.h&gt;</code><br/>
</div>
<p><a href="adapter_trimming_8h_source.html">Go to the source code of this file.</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structseqan_1_1_adapter_scoring_matrix.html">seqan::AdapterScoringMatrix</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Struct used to define a new custom scoring matrix.  <a href="structseqan_1_1_adapter_scoring_matrix.html#details">More...</a><br/></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structseqan_1_1_scoring_matrix_data___3_01int_00_01_dna5_00_01_adapter_scoring_matrix_01_4.html">seqan::ScoringMatrixData_&lt; int, Dna5, AdapterScoringMatrix &gt;</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Struct containing data for the <a class="el" href="structseqan_1_1_adapter_scoring_matrix.html" title="Struct used to define a new custom scoring matrix.">seqan::AdapterScoringMatrix</a> custom scoring matrix. Matches score 1, mismatches -1 and matches against N with 0.  <a href="structseqan_1_1_scoring_matrix_data___3_01int_00_01_dna5_00_01_adapter_scoring_matrix_01_4.html#details">More...</a><br/></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_mode.html">Mode</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A struct encapsulating information about the match algorithm.  <a href="struct_mode.html#details">More...</a><br/></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_auto.html">Auto</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Tagging struct for the automatic match algorithm.  <a href="struct_auto.html#details">More...</a><br/></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_user.html">User</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Tagging struct representing the the match algorithm working with values supplied by the user. Saves those values as members.  <a href="struct_user.html#details">More...</a><br/></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_adapter_trimming_stats.html">AdapterTrimmingStats</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Struct to hold information about certain adapter trimming statistics.  <a href="struct_adapter_trimming_stats.html#details">More...</a><br/></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_s_t_r_i_n_g___r_e_v_e_r_s_e___c_o_m_p_l_e_m_e_n_t.html">STRING_REVERSE_COMPLEMENT&lt; TValue &gt;</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A metafunction which constructs a ModifiedString type for an Alphabet.  <a href="struct_s_t_r_i_n_g___r_e_v_e_r_s_e___c_o_m_p_l_e_m_e_n_t.html#details">More...</a><br/></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="typedef-members"></a>
Typedefs</h2></td></tr>
<tr class="memitem:ab719b1a56b7f138316a70a055c8ae54e"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab719b1a56b7f138316a70a055c8ae54e"></a>
typedef Score&lt; int, <br class="typebreak"/>
ScoreMatrix&lt; Dna5, <br class="typebreak"/>
<a class="el" href="structseqan_1_1_adapter_scoring_matrix.html">AdapterScoringMatrix</a> &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>TScore</b></td></tr>
<tr class="separator:ab719b1a56b7f138316a70a055c8ae54e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:a690b4222b093c2260ab0522bff30049f"><td class="memTemplParams" colspan="2">template&lt;typename TSeq1 , typename TSeq2 , bool TTop, bool TLeft, bool TRight, bool TBottom&gt; </td></tr>
<tr class="memitem:a690b4222b093c2260ab0522bff30049f"><td class="memTemplItemLeft" align="right" valign="top">seqan::Pair&lt; unsigned, <br class="typebreak"/>
seqan::Align&lt; TSeq1 &gt; &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="adapter_trimming_8h.html#a690b4222b093c2260ab0522bff30049f">alignPair</a> (TSeq1 &amp;seq1, TSeq2 &amp;seq2, const seqan::AlignConfig&lt; TTop, TLeft, TRight, TBottom &gt; &amp;config, bool band=false)</td></tr>
<tr class="memdesc:a690b4222b093c2260ab0522bff30049f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Align two sequences, returning the score and align object of the alignment.  <a href="#a690b4222b093c2260ab0522bff30049f">More...</a><br/></td></tr>
<tr class="separator:a690b4222b093c2260ab0522bff30049f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3bc646d4dc99acf0ae165cf082e217cb"><td class="memTemplParams" colspan="2">template&lt;typename TRow &gt; </td></tr>
<tr class="memitem:a3bc646d4dc99acf0ae165cf082e217cb"><td class="memTemplItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="adapter_trimming_8h.html#a3bc646d4dc99acf0ae165cf082e217cb">countTotalGaps</a> (TRow &amp;row)</td></tr>
<tr class="memdesc:a3bc646d4dc99acf0ae165cf082e217cb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the total number of gaps in a row object.  <a href="#a3bc646d4dc99acf0ae165cf082e217cb">More...</a><br/></td></tr>
<tr class="separator:a3bc646d4dc99acf0ae165cf082e217cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af28bd342d3520fd7fe6e298e0271e288"><td class="memTemplParams" colspan="2">template&lt;typename TAlign &gt; </td></tr>
<tr class="memitem:af28bd342d3520fd7fe6e298e0271e288"><td class="memTemplItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="adapter_trimming_8h.html#af28bd342d3520fd7fe6e298e0271e288">getOverlap</a> (TAlign &amp;align)</td></tr>
<tr class="memdesc:af28bd342d3520fd7fe6e298e0271e288"><td class="mdescLeft">&#160;</td><td class="mdescRight">Determines the overlap between two (free-shift) aligned gapless sequences.  <a href="#af28bd342d3520fd7fe6e298e0271e288">More...</a><br/></td></tr>
<tr class="separator:af28bd342d3520fd7fe6e298e0271e288"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8427805bc99660a9476176b62061f45f"><td class="memTemplParams" colspan="2">template&lt;typename TAlign &gt; </td></tr>
<tr class="memitem:a8427805bc99660a9476176b62061f45f"><td class="memTemplItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="adapter_trimming_8h.html#a8427805bc99660a9476176b62061f45f">getInsertSize</a> (TAlign &amp;align)</td></tr>
<tr class="memdesc:a8427805bc99660a9476176b62061f45f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given an alignment with two overlapping (forward and reverse) sequences, this function determines the actual size of the insert they cover.  <a href="#a8427805bc99660a9476176b62061f45f">More...</a><br/></td></tr>
<tr class="separator:a8427805bc99660a9476176b62061f45f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a502e2eb7a85e1f67dd637e94f1600524"><td class="memTemplParams" colspan="2">template&lt;typename TSeq &gt; </td></tr>
<tr class="memitem:a502e2eb7a85e1f67dd637e94f1600524"><td class="memTemplItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="adapter_trimming_8h.html#a502e2eb7a85e1f67dd637e94f1600524">stripPair</a> (TSeq &amp;seq1, TSeq &amp;seq2)</td></tr>
<tr class="memdesc:a502e2eb7a85e1f67dd637e94f1600524"><td class="mdescLeft">&#160;</td><td class="mdescRight">Removes adapter contamination from paired-end reads.  <a href="#a502e2eb7a85e1f67dd637e94f1600524">More...</a><br/></td></tr>
<tr class="separator:a502e2eb7a85e1f67dd637e94f1600524"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aeddd7c1ce0243088be7cde1b4cb5825c"><td class="memTemplParams" colspan="2">template&lt;typename TSeq &gt; </td></tr>
<tr class="memitem:aeddd7c1ce0243088be7cde1b4cb5825c"><td class="memTemplItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="adapter_trimming_8h.html#aeddd7c1ce0243088be7cde1b4cb5825c">stripPair</a> (TSeq &amp;seq1, TSeq &amp;adapter1, TSeq &amp;seq2, TSeq &amp;adapter2)</td></tr>
<tr class="memdesc:aeddd7c1ce0243088be7cde1b4cb5825c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Removes adapter contamination from paired-end reads with adapter information.  <a href="#aeddd7c1ce0243088be7cde1b4cb5825c">More...</a><br/></td></tr>
<tr class="separator:aeddd7c1ce0243088be7cde1b4cb5825c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a390ded6489d24d870c5b63d57b394ad9"><td class="memTemplParams" colspan="2">template&lt;typename TSeq &gt; </td></tr>
<tr class="memitem:a390ded6489d24d870c5b63d57b394ad9"><td class="memTemplItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="adapter_trimming_8h.html#a390ded6489d24d870c5b63d57b394ad9">stripPairBatch</a> (seqan::StringSet&lt; TSeq &gt; &amp;set1, seqan::StringSet&lt; TSeq &gt; &amp;set2, <a class="el" href="struct_adapter_trimming_stats.html">AdapterTrimmingStats</a> &amp;stats)</td></tr>
<tr class="memdesc:a390ded6489d24d870c5b63d57b394ad9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Removes adapter contamination from paired-end reads.  <a href="#a390ded6489d24d870c5b63d57b394ad9">More...</a><br/></td></tr>
<tr class="separator:a390ded6489d24d870c5b63d57b394ad9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5c6fd3864a1835e6d7120555021326ee"><td class="memTemplParams" colspan="2">template&lt;typename TSeq , typename TAdapter &gt; </td></tr>
<tr class="memitem:a5c6fd3864a1835e6d7120555021326ee"><td class="memTemplItemLeft" align="right" valign="top">seqan::Pair&lt; unsigned, <br class="typebreak"/>
seqan::Align&lt; TSeq &gt; &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="adapter_trimming_8h.html#a5c6fd3864a1835e6d7120555021326ee">alignAdapter</a> (TSeq &amp;seq, TAdapter &amp;adapter)</td></tr>
<tr class="memdesc:a5c6fd3864a1835e6d7120555021326ee"><td class="mdescLeft">&#160;</td><td class="mdescRight">Aligns a sequence to an adapter.  <a href="#a5c6fd3864a1835e6d7120555021326ee">More...</a><br/></td></tr>
<tr class="separator:a5c6fd3864a1835e6d7120555021326ee"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8d41cfe5174b6faee428503c8fd0fa45"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="adapter_trimming_8h.html#a8d41cfe5174b6faee428503c8fd0fa45">isMatch</a> (int overlap, int mismatches, const <a class="el" href="struct_auto.html">Auto</a> &amp;)</td></tr>
<tr class="memdesc:a8d41cfe5174b6faee428503c8fd0fa45"><td class="mdescLeft">&#160;</td><td class="mdescRight">Checks if a overlap of an alignment is accepted, based on mismatches and the length of the overlap.  <a href="#a8d41cfe5174b6faee428503c8fd0fa45">More...</a><br/></td></tr>
<tr class="separator:a8d41cfe5174b6faee428503c8fd0fa45"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acd23c187993f2c93b6c9177d7df4563f"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="adapter_trimming_8h.html#acd23c187993f2c93b6c9177d7df4563f">isMatch</a> (int overlap, int mismatches, const <a class="el" href="struct_user.html">User</a> &amp;userOptions)</td></tr>
<tr class="memdesc:acd23c187993f2c93b6c9177d7df4563f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Checks if a overlap of an alignment is accepted, based on mismatches and the length of the overlap.  <a href="#acd23c187993f2c93b6c9177d7df4563f">More...</a><br/></td></tr>
<tr class="separator:acd23c187993f2c93b6c9177d7df4563f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7f0ed5b254d423dabfb235c79355242d"><td class="memTemplParams" colspan="2">template&lt;typename TSeq , typename TAdapter , typename TSpec &gt; </td></tr>
<tr class="memitem:a7f0ed5b254d423dabfb235c79355242d"><td class="memTemplItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="adapter_trimming_8h.html#a7f0ed5b254d423dabfb235c79355242d">stripAdapter</a> (TSeq &amp;seq, TAdapter &amp;adapter, TSpec &amp;spec)</td></tr>
<tr class="memdesc:a7f0ed5b254d423dabfb235c79355242d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Remove adapter sequence from a sequence.  <a href="#a7f0ed5b254d423dabfb235c79355242d">More...</a><br/></td></tr>
<tr class="separator:a7f0ed5b254d423dabfb235c79355242d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3c85c17ed0ce3241d328d9b609d88789"><td class="memTemplParams" colspan="2">template&lt;typename TSeq , typename TAdapter , typename TSpec &gt; </td></tr>
<tr class="memitem:a3c85c17ed0ce3241d328d9b609d88789"><td class="memTemplItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="adapter_trimming_8h.html#a3c85c17ed0ce3241d328d9b609d88789">stripAdapterBatch</a> (seqan::StringSet&lt; TSeq &gt; &amp;set, TAdapter &amp;adapter, TSpec const &amp;spec, <a class="el" href="struct_adapter_trimming_stats.html">AdapterTrimmingStats</a> &amp;stats, bool reverse=false)</td></tr>
<tr class="memdesc:a3c85c17ed0ce3241d328d9b609d88789"><td class="mdescLeft">&#160;</td><td class="mdescRight">Remove adapter sequence from a set of sequences.  <a href="#a3c85c17ed0ce3241d328d9b609d88789">More...</a><br/></td></tr>
<tr class="separator:a3c85c17ed0ce3241d328d9b609d88789"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a388d9f7c52eada005728f2394f2bb7d2"><td class="memTemplParams" colspan="2">template&lt;typename TSeq , typename TSpec &gt; </td></tr>
<tr class="memitem:a388d9f7c52eada005728f2394f2bb7d2"><td class="memTemplItemLeft" align="right" valign="top">unsigned&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="adapter_trimming_8h.html#a388d9f7c52eada005728f2394f2bb7d2">stripReverseAdapterBatch</a> (seqan::StringSet&lt; TSeq &gt; &amp;set, TSeq &amp;adapter, TSpec const &amp;spec, <a class="el" href="struct_adapter_trimming_stats.html">AdapterTrimmingStats</a> &amp;stats)</td></tr>
<tr class="memdesc:a388d9f7c52eada005728f2394f2bb7d2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Simple interface to align the reverse complement of an adapter to a batch of sequences. to a set of reads and remove significant matches.  <a href="#a388d9f7c52eada005728f2394f2bb7d2">More...</a><br/></td></tr>
<tr class="separator:a388d9f7c52eada005728f2394f2bb7d2"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Contains the functions for adapter trimming. </p>
</div><h2 class="groupheader">Function Documentation</h2>
<a class="anchor" id="a5c6fd3864a1835e6d7120555021326ee"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename TSeq , typename TAdapter &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">seqan::Pair&lt;unsigned, seqan::Align&lt;TSeq&gt; &gt; alignAdapter </td>
          <td>(</td>
          <td class="paramtype">TSeq &amp;&#160;</td>
          <td class="paramname"><em>seq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">TAdapter &amp;&#160;</td>
          <td class="paramname"><em>adapter</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Aligns a sequence to an adapter. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">seq</td><td>The sequence which adapter contamination shall be removed. </td></tr>
    <tr><td class="paramname">adapter</td><td>The adapter that might contaminate the sequence. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A pair of unsigned ints containing the score of the alignment and the alignment object. </dd></dl>

</div>
</div>
<a class="anchor" id="a690b4222b093c2260ab0522bff30049f"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename TSeq1 , typename TSeq2 , bool TTop, bool TLeft, bool TRight, bool TBottom&gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">seqan::Pair&lt;unsigned, seqan::Align&lt;TSeq1&gt; &gt; alignPair </td>
          <td>(</td>
          <td class="paramtype">TSeq1 &amp;&#160;</td>
          <td class="paramname"><em>seq1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">TSeq2 &amp;&#160;</td>
          <td class="paramname"><em>seq2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const seqan::AlignConfig&lt; TTop, TLeft, TRight, TBottom &gt; &amp;&#160;</td>
          <td class="paramname"><em>config</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>band</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Align two sequences, returning the score and align object of the alignment. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">seq1</td><td>The first sequence of the alignment. </td></tr>
    <tr><td class="paramname">seq2</td><td>The second sequence of the alignment. </td></tr>
    <tr><td class="paramname">config</td><td>The alignment configuration used by the Seqan's globalAlignment method. </td></tr>
    <tr><td class="paramname">band</td><td>Whether the alignment's lower diagonal should be banded at -2. (Two diagonals below the main diagonal.) Useful to speed up computation for 5'-3' overlap alignments. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section remark"><dt>Remarks</dt><dd>The main intended uses are AlignConfig&lt;true, false, true, false&gt; with band = true and AlignConfig&lt;true, true, true, true&gt; with band = false. </dd>
<dd>
We use a custom scoring scheme which scores 1 for match, -1 for mismatch, 0 for match with N. </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>A pair containing the score of the alignment and the aling object. </dd></dl>

</div>
</div>
<a class="anchor" id="a3bc646d4dc99acf0ae165cf082e217cb"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename TRow &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">unsigned countTotalGaps </td>
          <td>(</td>
          <td class="paramtype">TRow &amp;&#160;</td>
          <td class="paramname"><em>row</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the total number of gaps in a row object. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">row</td><td>The gapped row object used by seqan alignments. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a8427805bc99660a9476176b62061f45f"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename TAlign &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">unsigned getInsertSize </td>
          <td>(</td>
          <td class="paramtype">TAlign &amp;&#160;</td>
          <td class="paramname"><em>align</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Given an alignment with two overlapping (forward and reverse) sequences, this function determines the actual size of the insert they cover. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">align</td><td>An alignment object with two aligned overlapping sequences. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section pre"><dt>Precondition</dt><dd>The aligned sequences must overlap and not contain any internal gaps. </dd></dl>
<dl class="section remark"><dt>Remarks</dt><dd>This method can reconstruct the insert perfectly, provided the alignment object represents the real alignment. There can be small discrepancies if indels occurred. </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>The insert size that was determined from the overlap alignment. </dd></dl>

</div>
</div>
<a class="anchor" id="af28bd342d3520fd7fe6e298e0271e288"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename TAlign &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">unsigned getOverlap </td>
          <td>(</td>
          <td class="paramtype">TAlign &amp;&#160;</td>
          <td class="paramname"><em>align</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Determines the overlap between two (free-shift) aligned gapless sequences. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">align</td><td>An alignment object with two aligned overlapping sequences. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section pre"><dt>Precondition</dt><dd>The aligned sequences must not contain any internal gaps. </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>The number of overlapping positions. </dd></dl>

</div>
</div>
<a class="anchor" id="a8d41cfe5174b6faee428503c8fd0fa45"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool isMatch </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>overlap</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>mismatches</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_auto.html">Auto</a> &amp;&#160;</td>
          <td class="paramname">&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Checks if a overlap of an alignment is accepted, based on mismatches and the length of the overlap. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">overlap</td><td>The number of overlapping positions in the overlap alignment. </td></tr>
    <tr><td class="paramname">mismatches</td><td>The number of allowed mismatches in the overlapping region. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section remark"><dt>Remarks</dt><dd>This method automatically uses a very simple heuristic to determine matches. </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>Bool indicating if the alignment is significant. </dd></dl>

</div>
</div>
<a class="anchor" id="acd23c187993f2c93b6c9177d7df4563f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool isMatch </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>overlap</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>mismatches</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_user.html">User</a> &amp;&#160;</td>
          <td class="paramname"><em>userOptions</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Checks if a overlap of an alignment is accepted, based on mismatches and the length of the overlap. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">overlap</td><td>The number of overlapping positions in the overlap alignment. </td></tr>
    <tr><td class="paramname">mismatches</td><td>The number of allowed mismatches in the overlapping region. </td></tr>
    <tr><td class="paramname">userOptions</td><td>Parameters specifying requirements for the overlap. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section remark"><dt>Remarks</dt><dd>This overload is used to let the user specify match requirements. </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>Bool indicating if the alignment is significant. </dd></dl>

</div>
</div>
<a class="anchor" id="a7f0ed5b254d423dabfb235c79355242d"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename TSeq , typename TAdapter , typename TSpec &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">unsigned stripAdapter </td>
          <td>(</td>
          <td class="paramtype">TSeq &amp;&#160;</td>
          <td class="paramname"><em>seq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">TAdapter &amp;&#160;</td>
          <td class="paramname"><em>adapter</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">TSpec &amp;&#160;</td>
          <td class="paramname"><em>spec</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Remove adapter sequence from a sequence. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">seq</td><td>The sequence whose adapter contamination should be removed. </td></tr>
    <tr><td class="paramname">adapter</td><td>The adapter sequence that might contaminate the sequence. </td></tr>
    <tr><td class="paramname">spec</td><td>The match algorithm that decides whether a match was significant. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The overlap of the sequence with the adapter. </dd></dl>

</div>
</div>
<a class="anchor" id="a3c85c17ed0ce3241d328d9b609d88789"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename TSeq , typename TAdapter , typename TSpec &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">unsigned stripAdapterBatch </td>
          <td>(</td>
          <td class="paramtype">seqan::StringSet&lt; TSeq &gt; &amp;&#160;</td>
          <td class="paramname"><em>set</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">TAdapter &amp;&#160;</td>
          <td class="paramname"><em>adapter</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">TSpec const &amp;&#160;</td>
          <td class="paramname"><em>spec</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_adapter_trimming_stats.html">AdapterTrimmingStats</a> &amp;&#160;</td>
          <td class="paramname"><em>stats</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>reverse</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Remove adapter sequence from a set of sequences. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">set</td><td>A StringSet of sequences whose adapter contaminations should be removed. </td></tr>
    <tr><td class="paramname">adapter</td><td>The adapter sequence that might contaminate the sequences. </td></tr>
    <tr><td class="paramname">spec</td><td>The match algorithm that decides whether a match was significant. </td></tr>
    <tr><td class="paramname">reverse</td><td>Indicates whether we align forward or reverse sequences against an adapter. Exists mainly to enable correct writing of the appropriate fields when gathering statistics. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section remark"><dt>Remarks</dt><dd>This method simply applies <a class="el" href="adapter_trimming_8h.html#a7f0ed5b254d423dabfb235c79355242d" title="Remove adapter sequence from a sequence.">stripAdapter</a> to all sequences in <b>set</b>. </dd>
<dd>
This method can operate concurrently with OpenMP. </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>The number of sequences that had some bases removed. </dd></dl>

</div>
</div>
<a class="anchor" id="a502e2eb7a85e1f67dd637e94f1600524"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename TSeq &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">unsigned stripPair </td>
          <td>(</td>
          <td class="paramtype">TSeq &amp;&#160;</td>
          <td class="paramname"><em>seq1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">TSeq &amp;&#160;</td>
          <td class="paramname"><em>seq2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Removes adapter contamination from paired-end reads. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">seq1</td><td>The forward read. </td></tr>
    <tr><td class="paramname">seq2</td><td>The backward read. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section remark"><dt>Remarks</dt><dd>This method is very accurate and does not need any knowledge of the specific adapter sequences contaminating the reads. </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>The determined actual insert size or 0 if the sequences don't overlap. </dd></dl>
<dl class="section remark"><dt>Remarks</dt><dd>The number of trimmed bases is len(seq) - insert, if the sequence was greater than the insert size. </dd></dl>

</div>
</div>
<a class="anchor" id="aeddd7c1ce0243088be7cde1b4cb5825c"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename TSeq &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">unsigned stripPair </td>
          <td>(</td>
          <td class="paramtype">TSeq &amp;&#160;</td>
          <td class="paramname"><em>seq1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">TSeq &amp;&#160;</td>
          <td class="paramname"><em>adapter1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">TSeq &amp;&#160;</td>
          <td class="paramname"><em>seq2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">TSeq &amp;&#160;</td>
          <td class="paramname"><em>adapter2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Removes adapter contamination from paired-end reads with adapter information. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">seq1</td><td>The forward read. </td></tr>
    <tr><td class="paramname">adapter1</td><td>The adapter that contaminates the forward read. </td></tr>
    <tr><td class="paramname">seq2</td><td>The backward read. </td></tr>
    <tr><td class="paramname">adapter2</td><td>The adapter whose reverse complement contaminates the backward read. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section remark"><dt>Remarks</dt><dd>This method can be less accurate than the overload which doesn't use any adapter sequence information, since it is more constrained in how it tries to overlap the sequences. </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>The determined actual insert size or 0 if the sequences don't overlap. </dd></dl>
<dl class="section remark"><dt>Remarks</dt><dd>The number of trimmed bases is len(seq) - insert, if the sequence was greater than the insert size. </dd></dl>

</div>
</div>
<a class="anchor" id="a390ded6489d24d870c5b63d57b394ad9"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename TSeq &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">unsigned stripPairBatch </td>
          <td>(</td>
          <td class="paramtype">seqan::StringSet&lt; TSeq &gt; &amp;&#160;</td>
          <td class="paramname"><em>set1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">seqan::StringSet&lt; TSeq &gt; &amp;&#160;</td>
          <td class="paramname"><em>set2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_adapter_trimming_stats.html">AdapterTrimmingStats</a> &amp;&#160;</td>
          <td class="paramname"><em>stats</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Removes adapter contamination from paired-end reads. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">seq1</td><td>The set of forward reads. </td></tr>
    <tr><td class="paramname">seq2</td><td>The set of backward reads. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section remark"><dt>Remarks</dt><dd>This method is very accurate and does not need any knowledge of the specific adapter sequences contaminating the reads. </dd>
<dd>
This method can operate concurrently with OpenMP. </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>The number of sequences that had some bases removed. </dd></dl>

</div>
</div>
<a class="anchor" id="a388d9f7c52eada005728f2394f2bb7d2"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename TSeq , typename TSpec &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">unsigned stripReverseAdapterBatch </td>
          <td>(</td>
          <td class="paramtype">seqan::StringSet&lt; TSeq &gt; &amp;&#160;</td>
          <td class="paramname"><em>set</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">TSeq &amp;&#160;</td>
          <td class="paramname"><em>adapter</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">TSpec const &amp;&#160;</td>
          <td class="paramname"><em>spec</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_adapter_trimming_stats.html">AdapterTrimmingStats</a> &amp;&#160;</td>
          <td class="paramname"><em>stats</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Simple interface to align the reverse complement of an adapter to a batch of sequences. to a set of reads and remove significant matches. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">set</td><td>A StringSet of sequences whose adapter contaminations should be removed. </td></tr>
    <tr><td class="paramname">adapter</td><td>The adapter sequence whose reverse complement might contaminate the sequences. </td></tr>
    <tr><td class="paramname">spec</td><td>The match algorithm that decides whether a match was significant. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section remark"><dt>Remarks</dt><dd>This interface can be used to detect contamination in reverse reads of paired-end reads. Those reads might read into the reverse complement of an adapter. </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>The number of sequences that had some bases removed. </dd></dl>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Thu May 23 2013 12:09:04 for SeqDPT by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.3.1
</small></address>
</body>
</html>
